main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -44,31 +50,575 @@ object Lexer {
+      case _ if keyword("override") => TokenKind.OverrideKeyword
+      case _ if keyword("pub") => TokenKind.PubKeyword
+      case _ if keyword("ref") => TokenKind.RefKeyword
+      case _ if keyword("region") => TokenKind.RegionKeyword
+      case _ if keyword("reify") => TokenKind.ReifyKeyword
+      case _ if keyword("reifyBool") => TokenKind.ReifyBoolKeyword
+      case _ if keyword("reifyEff") => TokenKind.ReifyEffKeyword
+      case _ if keyword("reifyType") => TokenKind.ReifyTypeKeyword
+      case _ if keyword("rel") => TokenKind.RelKeyword
+      case _ if keyword("sealed") => TokenKind.SealedKeyword
+      case _ if keyword("set") => TokenKind.SetKeyword
+      case _ if keyword("spawn") => TokenKind.SpawnKeyword
+      case _ if keyword("Static") => TokenKind.StaticKeyword
+      case _ if keyword("true") => TokenKind.TrueKeyword
+      case _ if keyword("type") => TokenKind.TypeKeyword
+      case _ if keyword("use") => TokenKind.UseKeyword
+      case _ if keyword("where") => TokenKind.WhereKeyword
+      case _ if keyword("with") => TokenKind.WithKeyword
+      case _ if keyword("discard") => TokenKind.DiscardKeyword
+      case _ if keyword("object") => TokenKind.ObjectKeyword
+      case _ if keyword("par") => TokenKind.ParKeyword
+      case _ if keyword("yield") => TokenKind.YieldKeyword
+      case _ if isMathNameChar(c) => mathName()
+      case _ if isGreekNameChar(c) => greekName()
+
+      // User defined operators
+      case _ if validUserOpTokens.contains(c) => {
+        val p = peek()
+        if (validUserOpTokens.contains(p)) {
+          userDefinedOp()
+        } else if (c == '-' && p.isDigit) {
+          number() // negative numbers
+        } else if (c == '$' && p.isLetter) {
+          builtIn()
+        } else {
+          validUserOpTokens.apply(c)
+        }
+      }
+      case c if c.isLetter => name(c.isUpper)
+      case c if c.isDigit => number()
+      case '\"' => string()
+      case '\'' => char()
+      case '`' => infixFunction()
+      case _ => TokenKind.Err(LexerErr.UnexpectedChar)
+    }
 
-  private def isAtEnd()(implicit s: State): Boolean = ???
+    addToken(kind)
+  }
+
+  // Adds a token by consuming the characters between start and current
+  private def addToken(k: TokenKind)(implicit s: State): Unit = {
+    val t = s.src.data.slice(s.start.offset, s.current.offset).mkString("")
+    s.tokens += Token(k, t, s.start.line, s.start.column)
+    s.start = new Position(s.current.line, s.current.column, s.current.offset)
+  }
+
+  // Checks whether the following substring matches a keyword. Note that *comparison includes current*
+  private def keyword(k: String)(implicit s: State): Boolean = {
+    // check if the keyword can appear before eof
+    if (s.current.offset + k.length > s.src.data.length) {
+      return false
+    }
+
+    val start = s.current.offset - 1
+    val matches = s.src.data.slice(start, start + k.length).sameElements(k.toCharArray)
```

Comment:
This looks slow and on a hot path. Assume that keywords can only be ASCII (not emojis) then just write an ordinary loop. Thus avoiding costly slice, toCharArray, and sameElements.

Reply:
I changed this to a while loop to avoid allocations. Now the lexer phase runs at roughly 1/4 of the current parser!
```
  Lexer                          114,3 ms (01,8%)
  Parser                         410,8 ms (06,4%)
```